### PR TITLE
TW-1302: Display no results found in search

### PR DESCRIPTION
**Pull request by nqhhdev**
### Ticket 
- #1302

### Resolved

https://github.com/linagora/twake-on-matrix/assets/99852347/8e2c615d-8f6e-45d3-b8ae-fbaffc5cd1ba

https://github.com/linagora/twake-on-matrix/assets/99852347/ca02f986-8f5d-4121-8d49-d4eb69c29522

### CODE REVIEW

File: lib/pages/chat_search/chat_search_view.dart [hunk 1]
```diff
@@ -85,7 +85,8 @@ class _ServerSearchView extends StatelessWidget {
         ValueListenableBuilder(
           valueListenable: serverSearchController.searchResultsNotifier,
           builder: (context, searchResults, child) {
-            final events = searchResults.searchResults
+            final events = (searchResults as PresentationServerSideSearch)
```
Comment (imGok) [inline lib/pages/chat_search/chat_search_view.dart:88]: IMO it's safer to make a condition before casting 
`if (searchResults is PresentationServerSideSearch) return child!`

File: lib/pages/search/server_search_view.dart [hunk 2]
```diff
@@ -24,74 +27,80 @@ class ServerSearchMessagesList extends StatelessWidget {
         valueListenable:
             searchController.serverSearchController.searchResultsNotifier,
         builder: (context, severSearchNotifier, child) {
```
Comment (imGok) [inline lib/pages/search/server_search_view.dart:29]:
```suggestion
        builder: (context, serverSearchNotifier, child) {
```

File: lib/pages/chat_search/chat_search_view.dart (hunk 1 above)
Comment (nqhhdev) [reply lib/pages/chat_search/chat_search_view.dart]: Yes, apply it in last commit

File: lib/presentation/model/search/presentation_server_side_empty_search.dart [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import 'package:fluffychat/presentation/model/search/presentation_server_side.dart';
+
+class PresentationServerSideEmptySearch extends PresentationServerSide {
```
Comment (sherlockvn) [inline lib/presentation/model/search/presentation_server_side_empty_search.dart:3]: you can create `PresentationServerSideEmptySearch`  extends the `PresentationServerSideSearch` with searchResults = []

File: lib/presentation/model/search/presentation_server_side.dart [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import 'package:equatable/equatable.dart';
+
+class PresentationServerSide with EquatableMixin {
```
Comment (sherlockvn) [inline lib/presentation/model/search/presentation_server_side.dart:3]: remove this class

File: lib/pages/search/server_search_controller.dart [hunk 5]
```diff
@@ -71,19 +74,24 @@ class ServerSearchController with SearchDebouncerMixin {
...
           updateNextBatch(success.nextBatch);
-          searchResultsNotifier.value = PresentationServerSideSearch(
-            searchResults: [
-              ...searchResultsNotifier.value.searchResults,
-              ...success.results ?? [],
-            ],
-          );
+          if (success.results?.isEmpty == true) {
+            searchResultsNotifier.value = PresentationServerSideEmptySearch();
+          } else {
```
Comment (sherlockvn) [inline lib/pages/search/server_search_controller.dart:79]: why don't you yield `PresentationServerSideEmptySearch` in interactor

File: lib/pages/search/server_search_controller.dart [hunk 6]
```diff
@@ -71,19 +74,24 @@ class ServerSearchController with SearchDebouncerMixin {
...
         }
       },
     );
   }
 
   void resetSearchResults() {
-    searchResultsNotifier.value = const PresentationServerSideSearch(
+    searchResultsNotifier.value = PresentationServerSideSearch(
       searchResults: [],
     );
```
Comment (sherlockvn) [inline lib/pages/search/server_search_controller.dart:96]: why not use `PresentationServerSideEmptySearch`, please use 1 class, make it less confused

File: lib/pages/search/server_search_controller.dart [hunk 7]
```diff
@@ -126,7 +134,9 @@ class ServerSearchController with SearchDebouncerMixin {
   void loadMore() {
     if (_searchCategories == null ||
         isLoadingMoreNotifier.value ||
-        searchResultsNotifier.value.searchResults.isEmpty ||
+        (searchResultsNotifier.value as PresentationServerSideSearch)
+            .searchResults
+            .isEmpty ||
```
[the remark this answers is not in the document]
Comment (sherlockvn) [reply lib/pages/search/server_search_controller.dart:139]: it will crash for sure

File: lib/pages/search/server_search_controller.dart (hunk 6 above)
Comment (nqhhdev) [reply lib/pages/search/server_search_controller.dart:96]: It's func reset search result, not case empty

**Comment by sherlockvn** [comment]
This case is not showing anything
image

**Review by sherlockvn — APPROVED** [review]

**Pull request merged**